### DISCUSSION TITLE
ROS2 Custom QoS

### DISCUSSION THREAD

**Discussion by yoelbassin** [foxglove/community#1010] [Q&A]
Hey!
My robotic system works in a lossy wi-fi environment where we must use our custom QoS profiles for controlled throughput and reliability. We want to use foxglove, or at least the foxglove bridge, but this is a really big issue for us. I will be happy to create needed pull requests for supporting this matter, but wanted to hear the community and the devs thoughts :)

**Comment by achim-k** [comment]
When subscribing, Foxglove bridge tries to [choose a QoS](https://github.com/foxglove/ros-foxglove-bridge/blob/dbb033e0be7ab3d5d3f22de6e11740eeef9bd5b1/ros2_foxglove_bridge/src/ros2_foxglove_bridge.cpp#L461-L525) that matches the available publishers, similar to how `ros2 topic echo` does it. If your publishers use your custom QoS profile, then this might just work. Or is there anything special about your QoS profile that causes issues?

**Reply by yoelbassin** [reply]
Well, the thing is that I want the foxglove bridge to be the only publisher between my GCS and the robot. I want the robot to get keep-alive messages from the GCS on a “best-effort” (don’t want re-transmissions since I’m on a very low bandwidth network) QoS, and status updates on a “reliable” (must be received by the robot, even if delayed) QoS

**Reply by achim-k** [reply]
Am I understanding you correctly in that you want to use foxglove bridge only to publish messages? Foxglove bridge uses a TCP based websocket connection which I don't think is very suited for an unreliable network where the robot would frequently disconnect. Maybe it's best if the robot periodically checks if the GCS is still reachable e.g. via a simple GET request

**Reply by yoelbassin** [reply]
Not quite - I use foxglove for visualization as well (I.e. subscriptions), but the challenge is in the publishing process. 

The architecture is a ground control unit where I have the bridge established, and multiple front-end clients talk with the bridge using web-sockets.

On the other side, I have my robot (composed of multiple ROS nodes).

So, I want to use the bridge as a middleman between the ground station and the ROS based robot - where the communication is over DDS (ROS), and that is exactly the point where my connection is unstable

**Reply by achim-k** [reply]
When choosing the QoS profile for the publisher, Foxglove bridge first looks if there are existing publishers on that topic and then chooses a matching one:
https://github.com/foxglove/ros-foxglove-bridge/blob/9bc30200524f55de6dab1a73c23a617218f16dcc/ros2_foxglove_bridge/src/ros2_foxglove_bridge.cpp#L610-L620

Does this not work for you? Or maybe there are no publishers and you would like to specify the default publishing QoS profile?

**Reply by yoelbassin** [reply]
> Or maybe there are no publishers and you would like to specify the default publishing QoS profile?

**Yes** that's exactly the issue! What do you think can be done in that case?

**Reply by achim-k** [reply]
I guess the easiest would be to create a simple script with a simple ROS node and publishers with the QoS profile that you want. The script does nothing else than instantiating the publishers and spinning, no messages are published. This should allow foxglove bridge to pick up the desired QoS profile for publishing.

Alternatively, feel free to submit a PR to the foxglove bridge repo which adds fallback QoS parameters for publishing.

**Reply by defunctzombie** [reply]
> Or maybe there are no publishers and you would like to specify the default publishing QoS profile?

@achim-k That logic is for advertising from the bridge - when subscribing to messages the bridge would still use the QoS of the publisher does it not? So for the above request of working with lossy networks it still seems like the ROS publisher side should have its QoS updated?

**Reply by achim-k** [reply]
Not sure if I understand correctly, but the request is about advertising from the bridge and not for subscribing topics.

**Reply by defunctzombie** [reply]
👍 I missed that part of the clarification in the thread (the original post did not specify this was about publishing behavior from the bridge or app).
